module/component/component.go

```diff
@@ -118,76 +117,119 @@ func RunComponent(ctx context.Context, componentFactory ComponentFactory, handle
+//
+// The ReadyDoneAware interface returned by this function can be used to synchronize with the
+// worker associated with workerID.
+// Caution: waiting on workers can easily create deadlocks. use with care
+type LookupFunc func(workerID string) (module.ReadyDoneAware, bool)
+
 // ComponentWorker represents a worker routine of a component.
-// It takes a SignalerContext which can be used to throw any irrecoverable errors it encouters,
+// It takes a SignalerContext which can be used to throw any irrecoverable errors it encounters,
 // as well as a ReadyFunc which must be called to signal that it is ready. The ComponentManager
 // waits until all workers have signaled that they are ready before closing its own Ready channel.
-type ComponentWorker func(ctx irrecoverable.SignalerContext, ready ReadyFunc)
+// It additionally takes a LookupFunc which allows the worker function to get a ReadyDoneAware
+// interface for any other worker routine. This provides a convenient interface for managing
+// inter worker dependencies.
+type ComponentWorker func(ctx irrecoverable.SignalerContext, ready ReadyFunc, lookup LookupFunc)
 
 // ComponentManagerBuilder provides a mechanism for building a ComponentManager
 type ComponentManagerBuilder interface {
 	// AddWorker adds a worker routine for the ComponentManager
-	AddWorker(ComponentWorker) ComponentManagerBuilder
+	AddWorker(string, ComponentWorker) ComponentManagerBuilder
 
 	// Build builds and returns a new ComponentManager instance
 	Build() *ComponentManager
+
+	// SetSerialStart configures the ComponentManager to start workers serially
+	// By default, workers are started in parallel and must have explicit dependency checks
+	// This allows using the order workers were added to manage dependencies implicitly
+	SetSerialStart(bool) ComponentManagerBuilder
 }
 
 type componentManagerBuilderImpl struct {
-	workers []ComponentWorker
+	mu              *sync.Mutex
+	built           bool
+	serialStart     bool
+	workers         []*worker
+	workersRegistry map[string]module.ReadyDoneAware
 }
 
 // NewComponentManagerBuilder returns a new ComponentManagerBuilder
 func NewComponentManagerBuilder() ComponentManagerBuilder {
-	return &componentManagerBuilderImpl{}
+	return &componentManagerBuilderImpl{
+		mu:              &sync.Mutex{},
+		workersRegistry: make(map[string]module.ReadyDoneAware),
+	}
 }
 
-func (c *componentManagerBuilderImpl) AddWorker(worker ComponentWorker) ComponentManagerBuilder {
+func (c *componentManagerBuilderImpl) AddWorker(id string, f ComponentWorker) ComponentManagerBuilder {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, exists := c.workersRegistry[id]; exists {
+		// worker ids must be unique. adding duplicate ids will break the lookupWorker function which
+		// may result in incorrect dependency ordering
+		panic(fmt.Sprintf("worker with id %s already exists", id))
+	}
+
+	worker := newWorker(id, f)
 	c.workers = append(c.workers, worker)
+	c.workersRegistry[id] = worker
```

Comment:
not strictly, but we need to preserve the order for serial startup, and also make it easy to look up by id. in theory we do an O(N) search for a particular worker since N will be small

Reply:
this has been removed